selfservice/strategy/code/strategy_login.go

```diff
@@ -0,0 +1,258 @@
+		}
+
+		addresse := []Address{
+			{
+				To:  identifier,
+				Via: identity.CodeAddressType(string(cred.IdentifierAddressType)),
+			},
+		}
+
+		// kratos only supports `email` identifiers at the moment with the code method
+		// this is validated in the identity validation step above
+		if err := s.deps.CodeSender().SendCode(ctx, f, i, addresse...); err != nil {
+			return errors.WithStack(err)
+		}
+
+		// sets the flow state to code sent
+		s.NextFlowState(f)
+
+		nodeData, err := json.Marshal(struct {
+			Identifier string `json:"identifier"`
+		}{
+			Identifier: p.Identifier,
+		})
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		if err := s.NewCodeUINodes(r, f, json.RawMessage(nodeData)); err != nil {
+			return err
+		}
+
+		f.Active = identity.CredentialsTypeCodeAuth
+		if err = s.deps.LoginFlowPersister().UpdateLoginFlow(ctx, f); err != nil {
+			return err
+		}
+
+		if x.IsJSONRequest(r) {
+			s.deps.Writer().Write(w, r, f)
+		} else {
+			http.Redirect(w, r, f.AppendTo(s.deps.Config().SelfServiceFlowLoginUI(ctx)).String(), http.StatusSeeOther)
+		}
+
+		// we return an error to the flow handler so that it does not continue execution of the hooks.
+		// we are not done with the login flow yet. The user needs to verify the code and then we need to persist the identity.
+		return errors.WithStack(flow.ErrCompletedByStrategy)
+	})
+
+	codeManager.RegisterVerifyCodeState(func(ctx context.Context, p *CodeStateManagerPayload) error {
+		s.deps.Logger().
+			WithSensitiveField("code", p.Code).
+			WithSensitiveField("identifier", p.Identifier).
+			Debug("Verifying login code")
+
+		// Step 1: Get the identity
+		i, _, err = s.getIdentity(ctx, p.Identifier)
+		if err != nil {
+			return err
+		}
+
+		loginCode, err := s.deps.LoginCodePersister().UseLoginCode(ctx, f.ID, i.ID, p.Code)
+		if err != nil {
+			if errors.Is(err, ErrCodeNotFound) {
+				return schema.NewLoginCodeInvalid()
+			}
+			return errors.WithStack(err)
+		}
+
+		i, err = s.deps.PrivilegedIdentityPool().GetIdentity(ctx, loginCode.IdentityID, identity.ExpandDefault)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		// Step 2: The code was correct
+		f.Active = identity.CredentialsTypeCodeAuth
+
+		// since nothing has errored yet, we can assume that the code is correct
+		// and we can update the login flow
+		s.NextFlowState(f)
+
+		if err := s.deps.LoginFlowPersister().UpdateLoginFlow(ctx, f); err != nil {
+			return errors.WithStack(err)
+		}
+
+		return nil
+	})
+
+	codeManager.RegisterAlreadyValidatedCodeState(func(ctx context.Context, _ *CodeStateManagerPayload) error {
+		return s.HandleLoginError(w, r, f, &p, errors.WithStack(schema.NewNoLoginStrategyResponsible()))
+	})
+
+	if err := codeManager.Run(r.Context()); err != nil {
+		if errors.Is(err, flow.ErrCompletedByStrategy) {
+			return nil, err
+		}
+		// the error is already handled by the registered code states
+		return i, s.HandleLoginError(w, r, f, &p, err)
+	}
+
+	return i, nil
```

Comment:
What if something fails in the callbacks above and we end up with an `i == nil`?